src/Stratis.Bitcoin/P2P/Peer/NetworkPeerClient.cs

```diff
@@ -0,0 +1,172 @@
+
+            this.Stream = this.tcpClient.Connected ? this.tcpClient.GetStream() : null;
+            this.ProcessingCompletion = new TaskCompletionSource<bool>();
+
+            this.writeLock = new AsyncLock();
+        }
+
+        /// <summary>
+        /// Connects the network client to the target server.
+        /// </summary>
+        /// <param name="endPoint">IP address and port to connect to.</param>
+        /// <param name="cancellation">Cancellation token that allows aborting the operation.</param>
+        /// <exception cref="OperationCanceledException">Thrown when the connection attempt was aborted.</exception>
+        public async Task ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation)
+        {
+            this.logger = this.loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.Id}-{endPoint}] ");
+            this.logger.LogTrace("({0}:'{1}')", nameof(endPoint), endPoint);
+
+            try
+            {
+                await Task.Run(async () => await this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port), cancellation);
+                this.Stream = this.tcpClient.GetStream();
+            }
+            catch (Exception e)
+            {
+                if (e is OperationCanceledException)
+                {
+                    this.logger.LogTrace("Connecting to '{0}' cancelled.", endPoint);
+                    this.logger.LogTrace("(-)[CANCELLED]");
+                }
+                else
+                {
+                    this.logger.LogDebug("Error connecting to '{0}', exception: {1}", endPoint, e.ToString());
+                    this.logger.LogTrace("(-)[UNHANDLED_EXCEPTION]");
+                }
+                throw;
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Sends data over the established connection.
+        /// </summary>
+        /// <param name="data">Data to send.</param>
+        /// <param name="cancellation">Cancellation token that allows aborting the operation.</param>
+        /// <exception cref="OperationCanceledException">Thrown when the connection was terminated or the cancellation token was cancelled.</exception>
+        public async Task SendAsync(byte[] data, CancellationToken cancellation)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(data), nameof(data.Length), data.Length);
+
+            using (await this.writeLock.LockAsync(cancellation))
+            {
+                if (this.Stream == null)
+                {
+                    this.logger.LogTrace("Connection has been terminated.");
+                    this.logger.LogTrace("(-)[NO_STREAM]");
+                    throw new OperationCanceledException();
+                }
+
+                try
+                {
+                    await this.Stream.WriteAsync(data, 0, data.Length, cancellation);
+                }
+                catch (Exception e)
```

Comment:
Same here why not catch the exception type?

Reply:
here it is little more complicated and i have a similar handling for IO and Cancel, so it seems better to me in this particular case to have it not separated, like discussed here https://stackoverflow.com/questions/136035/catch-multiple-exceptions-at-once

so here i would like to keep it as it is

Reply:
Yeah fair enough I didn't think about same logic scenarios